### ISSUE TITLE
Backoffice order blocked at 'Address' stage

### ISSUE THREAD

**Issue by lbwright22** [openfoodfoundation/openfoodnetwork#7337]
## Description

Create order in back office for enterprise 201078 using OC 207544 (also tested for OC 207889)
Unable to complete the order placement.
Can not access 'Payments' tab
Order remains with status = 'Address'

Product stock is overridden in inventory but inventory settings do not necessitate products being in the inventory for them to be added to an order cycle.
No tag rules
Products added to backoffice orders are in the corresponding OC and are in stock in inventory.
Enterprise has delivery shipping methods set up.

Tested another enterprise and back office order placement works as predicted.

## Expected Behavior

Create order in back office for enterprise 201078 using OC 207544 (also tested for OC 207889)
Add product in 'Order Details'
Visit 'Customer Details' and add address
Visit 'Payments' and add a payment
complete the order

## Actual Behaviour

Create order in back office for enterprise 201078 using OC 207544 (also tested for OC 207889)
Add product in 'Order Details'
Visit 'Customer details' and add address
click update
click on 'payments' to finalise order placement. 
Taken back to 'customer details'

Revisit incomplete order and edit.
Products removed from 'Order Details' section. Fees remain only.

## Steps to Reproduce

1. Create order in back office for enterprise 201078 using OC 207544 (also tested for OC 207889)
2. Add product in 'Order Details'
3. Visit 'Customer details' and add address
4. click update
5. click on 'payments' to finalise order placement. 
6. Taken back to 'customer details'

1. Revisit incomplete order and edit.
2. Products removed from 'Order Details' section. 
3. Observe Fees remain only.

## Severity

bug s2
## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

**Comment by filipefurtad0** [comment]
Thanks for reporting this @lbwright22 .

I first observed this yesterday while testing this PR#6924. I second your observation:

- As admin (pic below, left): The **shipping section** is missing from the Backoffice: It usually contains the **shipping number**, the **shipping state** and the **line items** to be shipped, **shipping method** and **tracking number**. 
- As customer (pic below, right): Attempting to visit a placed order from which the shipping section is missing -> snail

![image](https://user-images.githubusercontent.com/49817236/113892824-51fc4600-97be-11eb-8262-b87afb9fbd8d.png)

Not sure what is the pattern for these orders. It both admins and customers. Depending on the extent, could be an s1... Investigating.

**Comment by filipefurtad0** [comment]
For these orders, it appears that shipments are missing from the `spree_shipments`, in the DB (staging-UK).

A corrupt order:

```

   id   |   number   | item_total | total  |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |              email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
--------+------------+------------+--------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+---------------+----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 716543 | R631265404 |     100.00 | 127.00 | complete |            27.00 |  208036 | 2021-04-01 17:08:53.839138 | 2021-04-01 17:16:53.907742 | 2021-04-01 17:08:56.497015 |          427847 |          427848 |        127.00 | shipped        | paid          | filipefurtado+kurkumac@gmail.com |                      |         201402 | GBP      |                 |         203946 |        7419 |               |               0.00 |                 0.00

```
`openfoodnetwork=> SELECT * FROM spree_shipments WHERE order_id='716543';`

```
 id | tracking | number | cost | shipped_at | order_id | address_id | created_at | updated_at | state | stock_location_id | included_tax_total | additional_tax_total | adjustment_total 
----+----------+--------+------+------------+----------+------------+------------+------------+-------+-------------------+--------------------+----------------------+------------------
(0 rows)
```

A "healthy" order:

```

   id   |   number   | item_total | total |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |              email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
--------+------------+------------+-------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+---------------+----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 717290 | R001760052 |      15.00 | 15.00 | complete |             0.00 |  208036 | 2021-04-07 15:18:22.607755 | 2021-04-07 15:18:58.546236 | 2021-04-07 15:18:58.368436 |          427886 |          427887 |          0.00 | pending        | balance_due   | filipefurtado+kurkumac@gmail.com |                      |         201402 | GBP      |                 |         203946 |        7419 |               |               0.00 |                 0.00

```

`openfoodnetwork=> SELECT * FROM spree_shipments WHERE order_id='717290';`

```
   id   | tracking |    number    | cost | shipped_at | order_id | address_id |         created_at         |         updated_at         |  state  | stock_location_id | included_tax_total | additional_tax_total | adjustment_total 
--------+----------+--------------+------+------------+----------+------------+----------------------------+----------------------------+---------+-------------------+--------------------+----------------------+------------------
 237251 |          | H87668074776 | 0.00 |            |   717290 |            | 2021-04-07 15:18:48.682889 | 2021-04-07 15:18:58.530957 | pending |                 1 |               0.00 |                 0.00 |             0.00
(1 row)

```

**Comment by Matt-Yorkley** [comment]
I think I was trying some things out on UK Staging the other day, there's a chance I might have been me that deleted the shipments there. Sorry! I'll take a look at restoring a backup. I think it's separate from this issue in UK Production though. Deploying PRs to staging that were not merged before the last release also can't be the cause of the issue in UK Prod, right?

**Comment by Matt-Yorkley** [comment]
Going back to this comment: https://github.com/openfoodfoundation/openfoodnetwork/pull/6924#issuecomment-814330180

I also manually replicated something like this the first time it came up. Switching the toggle and restarting unicorn made the shipment section re-appear on the order I was looking at.

**Comment by filipefurtad0** [comment]
Thanks @Matt-Yorkley an no worries on that - restoring the orders might be an idea, deleting them as well would be fine.
> Deploying PRs to staging that were not merged before the last release can't be the cause of the issue in UK Prod, right?
Right :+1: It all matched with the time of the deployment of the patch, so for a moment I thought PR #6924  was in it (but it's not).

We should probably remove the default setting to enable unit prices (or features which are in development) from **staging environments** - I think  this was a conclusion from our meeting on this matter (cant' find the notes now).

**Comment by andrewpbrett** [comment]
I was able to reproduce this issue by downloading the UK database and trying to place an order containing one of the subscription coffee products. 

I'm not all that familiar with the inventory parts of the code base, but I was able to track this down and it was happening because it's saying there's no stock available for the item. 

In the database dump that I have, that OrderCycle has the following products available: 

```
["- New Dawn Traders / Yallah - Coffee", "- New Dawn Traders / Yallah - Coffee", "- New Dawn Traders / Yallah - Coffee", "- New Dawn Traders / Yallah - Coffee Subscription", "- New Dawn Traders / Yallah - Coffee Subscription", "- New Dawn Traders / Yallah - Coffee Subscription", "- New Dawn Traders / Yallah - Coffee Subscription", "- New Dawn Traders / Arhuaco - Panela", "- New Dawn Traders / Arhuaco - Panela", "- New Dawn Traders / Arhuaco - Panela", "- New Dawn Traders / Yallah - Coffee", "- New Dawn Traders / Yallah - Coffee", "- Gorreana - Green Tea Hysson", "- Gorreana - Green Tea Hysson", "- Gorreana - Black Tea Pekoe", "- Gorreana - Black Tea Pekoe", "- Atlantic Brewery - Azores Pale Ale -  4.2%", "- Atlantic Brewery - Azores Pale Ale -  4.2%"] 
```

But when I look at the inventory for New Dawn, none of those products appear. Again, this might be me misunderstanding what @lbwright22 mentions about:

>Product stock is overridden in inventory but inventory settings do not necessitate products being in the inventory for them to be added to an order cycle.
and
>Products added to backoffice orders are in the corresponding OC and are in stock in inventory.

But from what I'm looking at, it does look like something isn't right with how inventory is working for this distributor. 

I also tested creating an order for another distributor/order cycle and was able to move the order to the payment state after entering customer details. 

We could probably downgrade this from an S2 given that it's apparently limited to just this shop and might be fixed by checking their setup, depending on what @lbwright22 thinks. 

In any case, though, we could look into making the error message more informative - right now, we expect that entering customer details should always advance the order state to payment, and don't indicate why if that doesn't happen.

**Comment by andrewpbrett** [comment]
Actually, it looks like quite a few of these products in the order cycle - everything supplied by New Dawn traders in fact - has 0 on hand in the product list, and does not show up in the inventory. 

When creating a backoffice order, these products show up in the dropdown search as being able to add to the order. I don't know if that's a recent change/regression, but the immediate workaround would be to either add the product to inventory or change its quantity on hand to be > 0.

**Comment by lbwright22** [comment]
Hi @andrewpbrett all of New Dawn Traders products are in the New dawn Traders Inventory with stock level over-rides. It just takes ages for them to load and show.  They are already in stock in the inventory. Sorry...
![newdawntradersinventory](https://user-images.githubusercontent.com/45104035/115210402-72010300-a0f6-11eb-8173-08fcd15f0f22.gif)
I have just tested creating a back office order. The issue remains.
The products are visible on the shop front. It is a problem with back office orders for this enterprise

**Comment by lin-d-hop** [comment]
This is an S2. It would be an S1 with a report from another user. No backoffice orders. I have successfully placed a backoffice order on a different OC for this user but not with this OC. It seems something specific to the OC but I can't figure out why that might be...

I just want to add to the description:

Normal Behaviour:
1. Create order in back office for enterprise 201078 using OC 207544 (also tested for OC 207889)
2. Add product in 'Order Details'
3. Visit 'Customer details' and add address
4. click update
5. **Add shipping method in 'Order Details'**

On an order with these products for this enterprise in this order cycle you do not have the option to add the Shipping Method. And the line items do not display correctly as Louise described.

In the below image NORMALLY i would see line items and the option to choose a Shipping Method. Why can't I choose a shipping method? I get blocked by Order State 'Address'. Why does this state transition occur?

![Screenshot from 2021-04-19 11-20-27](https://user-images.githubusercontent.com/8669848/115221231-5b13de00-a101-11eb-9b06-80a02a027483.png)

**Comment by andrewpbrett** [comment]
Thanks @lin-d-hop and @lbwright22 for clarifying. 

I'm going to put some notes here on what I've discovered so far, since I'm scratching my head and might need to explain it to a [rubber duck](https://en.wikipedia.org/wiki/Rubber_duck_debugging).

The first-level reason that we can't access the payments screen is that we [assume](https://github.com/openfoodfoundation/openfoodnetwork/blob/aca19edd3b49fe333f50d2219e0f1a4e30ac377c/app/controllers/spree/admin/payments_controller.rb#L132) that filling in customer details will transition the order from the `address` to `payment` state. This isn't happening in this case. 

Okay, so why isn't it happening? We call `OrderWorkflow.new(@order).complete` after the customer details are filled in. This attempts to move the order to the payment state. In the process it checks how many of each line item's variant is on hand:

https://github.com/openfoodfoundation/openfoodnetwork/blob/aca19edd3b49fe333f50d2219e0f1a4e30ac377c/app/models/concerns/variant_stock.rb#L91-L101

In this case, `total_on_hand` is returning 0. 

There are actually four different places we define a method called `total_on_hand`. The one in `ScopeVariantToHub` looks for the presence of a VariantOverride; the others do not. When I put a debugger in each of them, I found that we call the `total_on_hand` method in `Spree::Variant`, which then uses the one in `Spree::Stock::Quantifier`, to determine if there's enough stock available. 

When creating the BO order, however, the dropdown uses the API::Variant::Serializer, which uses the `total_on_hand` method in `ScopeVariantToHub`. 

So basically, the variant override is being ignored here. I can't imagine that this we changed something such that suddenly all variant overrides are being ignored, or we'd be hearing about it from more than just this shop. 

I'm not all that familiar with the inventory/variant override parts of the code base (although I'm becoming more familiar by the hour...). I'm going to hit publish on this comment in case someone else wants to pick up the trail from here but I'll also continue to work on it.

**Comment by Matt-Yorkley** [comment]
:duck::100:

**Comment by Matt-Yorkley** [comment]
Inventory / VariantOverrides / "scoping" variants is crazy and head-scratching is 100% expected :ok_hand: 

So to continue the rubber-ducking; I think when that dropdown looks for variants, it's hitting `Admin::VariantsController#search`:
https://github.com/openfoodfoundation/openfoodnetwork/blob/479897fcbd9506980b42ab55a8c0988a58e95b5d/app/controllers/spree/admin/variants_controller.rb#L56-L60

Which essentially finds variants and then in theory "overrides" them here before returning the results:

https://github.com/openfoodfoundation/openfoodnetwork/blob/479897fcbd9506980b42ab55a8c0988a58e95b5d/lib/open_food_network/scope_variants_for_search.rb#L69-L74

The override process is a bit crazy but essentially it overwrites the attributes of the `Variant` object in memory with a different set of attributes from a corresponding `VariantOverride`, so by the time the `Api::Variant::Serializer` asks what the object's `on_hand` count is, it's basically not the same object. In theory it will respond with the stock level of the variant override and not the base variant... :grimacing:
